site-src/geps/gep-746.md

```diff
@@ -0,0 +1,265 @@
+
+## Introduction
+
+TLS Certificate references on HTTPRoute have always been a confusing part of the
+Gateway API. In the v1alpha2 release, we should consider removing this feature
+while we still can. This GEP proposes an alternative that is simpler to work
+with and understand, while also leaving sufficient room to enable all the same
+capabilities that certificate references on HTTPRoute enabled.
+
+### Attaching TLS Certificates with Routes is Confusing
+One of the most confusing parts of the Gateway API is how certificates can be
+attached to Routes. There are a variety of different factors that lead to
+confusion here:
+
+* It can be natural to assume that a certificate attached to a Route only
+  applies to that Route. In reality, it applies to the entire listener(s)
+  associated with that Route.
+* This means that a Route can affect any other Routes attached to the same
+  Gateway Listener. By attaching a Route to a Gateway Listener, you’re
+  implicitly trusting all other Routes attached to that Gateway Listener.
+* When multiple Routes specify a certificate for the same Listener, it’s
+  possible that they will conflict and create more confusion.
+
+### Why We Did It
+To understand how we ended up with the ability to attach TLS certificates with
+Routes, it’s helpful to look at the use cases for this capability:
+
+1. Some users want Route owners to be able to attach arbitrary domains and certs
+   to a Gateway listener.
+   [#103](https://github.com/kubernetes-sigs/gateway-api/issues/103)
+1. Some users want Route owners to control certs for their applications.
+
+### Alternative Solutions
+
+#### 1. Automation with tools like Cert-Manager
+When automation is acceptable, the first use case is entirely possible with
+tools like cert-manager that can watch Routes, generate certs for them, and
+attach them to a Gateway.
+
+#### 2. Cross Namespace Cert Direct References from Gateways
+With the already established ReferencePolicy concept, we have established a safe
+way to reference resources across namespaces. Although this would require some
+coordination between Gateway and App owners, it would enable App owners to
+retain full control of the certs used by their app without the extra confusion
+that certs in HTTPRoute have led to.
+
+### Enabling Self-Service Certificate Attachment for App Owners
+Although this dramatically simplifies the API, it does not completely replace
+the functionality that certs attached to HTTPRoutes enabled. Most notably, it
+would be difficult to attach arbitrary self-provided certificates to a Gateway
+listener without requiring manual changes from a Gateway admin.
+
+There are a couple potential solutions here:
+
+#### 1. Implement a selector for cert references instead of direct references
+Although the simplicity of this approach is nice, it ends up with many of the
+same problems as certificates attached to Routes have and feels inconsistent
+with how Routes attach to Gateways.
```

Comment:
I think we're going to be limited to a cert per domain, or maybe a cert per domain per gateway. If you had 1000 httproutes with their own cert, would they also have unique hostnames? The end result in any case is the same, these certs need to be attached to a Gateway. The question is how we do that. It could be a direct ref from Route -> Cert, or as proposed here, a direct ref from Gateway Listener -> Cert.

I think automation could _potentially_ help here. A controller could be configured to trust certs for *.example.com from foo namespace, and then could automatically attach them to a Gateway specified by some kind of label. That seems like it could be slightly easier than either Gateway or Route attachment when dealing with 1000+ domains. Though hopefully at that scale, cert generation can be automated with a tool like cert-manager.

Any way we do this, we'll need some kind of link between a Gateway and a cert. HTTPRoute provided one potential way to make that link, but I'm not sure it provided any significant scalability/usability improvements.

Reply:
> I think we're going to be limited to a cert per domain, or maybe a cert per domain per gateway. 

I might be missing some nuance.  Should "domain" be read as "hostname" here?

> If you had 1000 httproutes with their own cert, would they also have unique hostnames?

Yes.  

> I think automation could _potentially_ help here. A controller could be configured to trust certs for *.example.com from foo namespace, and then could automatically attach them to a Gateway specified by some kind of label. That seems like it could be slightly easier than either Gateway or Route attachment when dealing with 1000+ domains.

I'm less concerned about automating the attachment and more concerned about fanout and the idea that the gateway would need 1000+ direct references to certificates.  I suppose a controller could batch updates to mitigate contention issues, but it would be a bit inconvenient (especially for humans) to work with a gateway definition that had to list out explicitly every linked certificate.

Reply:
Oh, I just realized that gateway's `spec.listener` has `// +kubebuilder:validation:MaxItems=64`, and each listener can have only one certificate reference.

Reply:
> I might be missing some nuance. Should "domain" be read as "hostname" here?

Yep, I should've just said "hostname"

> I just realized that gateway's spec.listener has // +kubebuilder:validation:MaxItems=64, and each listener can have only one certificate reference.

There's some nuance here in that a single resource may be linked per listener, that resource could potentially contain more than one cert (not saying that's a good idea though).

I think one of the problems with certs being attached to Routes was that there was a bit too much magic going on. It's difficult/impossible to see which certs are actually attached to a given Gateway listener. Although certs appeared to be attached to Routes, they were being attached to the listeners attached to those Routes, and that was just being hidden. As an implementation you'd still need to have some way to represent that many certs being attached to your listeners. I'd rather increase the max number of listeners and/or turn certificate refs into a list because I think those changes would be clearer and easier for users to understand. 

> more concerned about fanout and the idea that the gateway would need 1000+ direct references to certificates

I agree that that's never going to be a good experience. Would Gateway merging help here? The idea that compatible Gateways can be merged together by implementations?

Reply:
> There's some nuance here in that a single resource may be linked per listener, that resource could potentially contain more than one cert (not saying that's a good idea though).

The godoc implies that the resource can have exactly one certificate+key pair:

> 	// CertificateRef is a reference to a Kubernetes object that contains a TLS
>	// certificate and private key. This certificate is used to establish a TLS
>	// handshake for requests that match the hostname of the associated listener.

> I agree that that's never going to be a good experience. Would Gateway merging help here? The idea that compatible Gateways can be merged together by implementations?

I was thinking through this possibility.  Gateway merging might be the best approach here.  Either the cluster admin could give users permission to create gateways in their own namespaces, which would allow users to attach routes and certificates within their own respective namespaces (or others' namespaces, but only if permitted by some referencepolicy); or the controller that automated certificate management could do some sort of sharding with up to 64 certificates per shard/gateway.  (I personally like the idea of allowing users to manage their own gateways, but I need to think through all the implications of that approach.)

Reply:
To be clear, I'm still on board with this GEP given how attaching a certificate to a route using the API now doesn't actually bind the certificate to the route, but the self-service use case still needs some clarification (not necessarily in this GEP though).

Reply:
Yeah I personally really like the idea of more smaller Gateways than a single massive one. I'm really hoping that pattern will work for most organizations. I think it's worth digging into this self service model a bit more, especially around the scale aspect. I'd primarily been thinking in terms of ~dozens of certs per Gateway, not 1000+. 

>To be clear, I'm still on board with this GEP given how attaching a certificate to a route using the API as it stands without this GEP doesn't actually bind the certificate to the route, but the self-service use case still needs some clarification (not necessarily in this GEP though).

Sounds good, and agree that we need a clearer understanding of how self-service could work. Maybe @maelvls or @jakexks would have some insights here around how we can build a more scalable API around this. 

Given the v1alpha2 timeline, I think I'm going to create a follow up issue to continue the discussion around this so we can get this change in.

Reply:
Created https://github.com/kubernetes-sigs/gateway-api/issues/763 to track this.